File: internal/cloudadmin/token.go
package cloudadmin

import (
	"context"
	"crypto/ed25519"
	"crypto/rand"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"errors"
	"regexp"
	"strings"
	"time"

	"github.com/bignormal/aera-admin/internal/rbac"
	"github.com/google/uuid"
)

const serviceTokenLifetime = 5 * time.Minute

var serviceIdentityPattern = regexp.MustCompile(`^[a-z][a-z0-9._-]{2,63}$`)

type tokenSource interface {
	Token(context.Context, *ActorContext) (string, error)
}

type ActorContext struct {
	AdminID          uuid.UUID
	Role             rbac.Role
	OperationID      *uuid.UUID
	ApprovalID       *uuid.UUID
	RequesterAdminID *uuid.UUID
}

type ed25519TokenSource struct {
	privateKey ed25519.PrivateKey
	issuer     string
	subject    string
	scopes     []string
	clock      func() time.Time
}

type serviceClaims struct {
	Issuer           string   `json:"iss"`
	Subject          string   `json:"sub"`
	Audience         string   `json:"aud"`
	Scopes           []string `json:"scope"`
	IssuedAt         int64    `json:"iat"`
	NotBefore        int64    `json:"nbf"`
	ExpiresAt        int64    `json:"exp"`
	JWTID            string   `json:"jti"`
	AdminID          string   `json:"admin_id,omitempty"`
	AdminRole        string   `json:"admin_role,omitempty"`
	OperationID      string   `json:"operation_id,omitempty"`
	ApprovalID       string   `json:"approval_id,omitempty"`
	RequesterAdminID string   `json:"requester_admin_id,omitempty"`
}

func parseEd25519PrivateKey(raw []byte) (ed25519.PrivateKey, error) {
	block, rest := pem.Decode(raw)
	if block == nil || block.Type != "PRIVATE KEY" || len(strings.TrimSpace(string(rest))) != 0 {
		return nil, errors.New("service JWT signing key must contain one PKCS8 private key")
	}
	parsed, err := x509.ParsePKCS8PrivateKey(block.Bytes)
	if err != nil {
		return nil, errors.New("service JWT signing key is invalid")
	}
	privateKey, ok := parsed.(ed25519.PrivateKey)
	if !ok || len(privateKey) != ed25519.PrivateKeySize {
		return nil, errors.New("service JWT signing key must be Ed25519")
	}
	return append(ed25519.PrivateKey(nil), privateKey...), nil
}

func newTokenSource(privateKey ed25519.PrivateKey, issuer, subject string, scopes []string, clock func() time.Time) (tokenSource, error) {
	if len(privateKey) != ed25519.PrivateKeySize || !serviceIdentityPattern.MatchString(issuer) ||
		!serviceIdentityPattern.MatchString(subject) || len(scopes) == 0 || clock == nil {
		return nil, errors.New("service JWT configuration is invalid")
	}
	return &ed25519TokenSource{
		privateKey: append(ed25519.PrivateKey(nil), privateKey...),
		issuer:     issuer,
		subject:    subject,
		scopes:     append([]string(nil), scopes...),
		clock:      clock,
	}, nil
}

func (source *ed25519TokenSource) Token(ctx context.Context, actor *ActorContext) (string, error) {
	if err := ctx.Err(); err != nil {
		return "", err
	}
	if !validActorContext(actor) {
		return "", errors.New("service JWT actor context is invalid")
	}
	identifier := make([]byte, 16)
	if _, err := rand.Read(identifier); err != nil {
		return "", errors.New("service JWT id could not be generated")
	}
	now := source.clock().UTC().Truncate(time.Second)
	claims := serviceClaims{
		Issuer: source.issuer, Subject: source.subject, Audience: "aera-cloud-admin",
		Scopes: append([]string(nil), source.scopes...), IssuedAt: now.Unix(),
		NotBefore: now.Add(-5 * time.Second).Unix(), ExpiresAt: now.Add(serviceTokenLifetime).Unix(),
		JWTID: base64.RawURLEncoding.EncodeToString(identifier),
	}
	if actor != nil {
		claims.AdminID = actor.AdminID.String()
		claims.AdminRole = string(actor.Role)
		if actor.OperationID != nil {
			claims.OperationID = actor.OperationID.String()
		}
		if actor.ApprovalID != nil {
			claims.ApprovalID = actor.ApprovalID.String()
			claims.RequesterAdminID = actor.RequesterAdminID.String()
		}
	}
	header, _ := json.Marshal(map[string]string{"alg": "EdDSA", "typ": "JWT"})
	body, err := json.Marshal(claims)
	if err != nil {
		return "", errors.New("service JWT claims could not be encoded")
	}
	unsigned := base64.RawURLEncoding.EncodeToString(header) + "." + base64.RawURLEncoding.EncodeToString(body)
	signature := ed25519.Sign(source.privateKey, []byte(unsigned))
	return unsigned + "." + base64.RawURLEncoding.EncodeToString(signature), nil
}

func validActorContext(actor *ActorContext) bool {
	if actor == nil {
		return true
	}
	if actor.AdminID == uuid.Nil || !actor.Role.Valid() ||
		(actor.OperationID != nil && *actor.OperationID == uuid.Nil) ||
		(actor.ApprovalID != nil && *actor.ApprovalID == uuid.Nil) ||
		(actor.RequesterAdminID != nil && *actor.RequesterAdminID == uuid.Nil) {
		return false
	}
	rollbackEvidence := actor.ApprovalID != nil || actor.RequesterAdminID != nil
	if rollbackEvidence {
		return actor.OperationID != nil && actor.ApprovalID != nil && actor.RequesterAdminID != nil &&
			*actor.RequesterAdminID != actor.AdminID
	}
	return true
}

File: src/platform-api/cloud/token.ts
import { createPrivateKey, type KeyObject, randomBytes, sign as signRaw } from 'node:crypto'

import { cloudAdminRoles } from '../../access/cloud-actor'

// 逐字段对齐 aera-admin/internal/cloudadmin/token.go 与
// aera-cloud/internal/adminapi/auth.go 的服务 JWT 契约：
// EdDSA(Ed25519) + aud=aera-cloud-admin + 5 分钟寿命 + 严格 claim 集合。
const serviceTokenLifetimeSeconds = 5 * 60
const notBeforeSkewSeconds = 5
const serviceJWTAudience = 'aera-cloud-admin'

const canonicalUUIDPattern = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/

export type CloudActorContext = {
  adminId: string
  role: string
  operationId?: string
  approvalId?: string
  requesterAdminId?: string
}

type ServiceClaims = {
  iss: string
  sub: string
  aud: string
  scope: readonly string[]
  iat: number
  nbf: number
  exp: number
  jti: string
  admin_id?: string
  admin_role?: string
  operation_id?: string
  approval_id?: string
  requester_admin_id?: string
}

export function isCanonicalUUID(value: string): boolean {
  return canonicalUUIDPattern.test(value)
}

export function parseEd25519PrivateKey(pem: string): KeyObject {
  let key: KeyObject
  try {
    key = createPrivateKey(pem)
  } catch {
    throw new Error('service JWT signing key is invalid')
  }
  if (key.asymmetricKeyType !== 'ed25519') {
    throw new Error('service JWT signing key must be Ed25519')
  }
  return key
}

export function validActorContext(actor: CloudActorContext | undefined): boolean {
  if (!actor) return true
  if (!isCanonicalUUID(actor.adminId)) return false
  if (!cloudAdminRoles.includes(actor.role as (typeof cloudAdminRoles)[number])) return false
  if (actor.operationId !== undefined && !isCanonicalUUID(actor.operationId)) return false
  if (actor.approvalId !== undefined && !isCanonicalUUID(actor.approvalId)) return false
  if (actor.requesterAdminId !== undefined && !isCanonicalUUID(actor.requesterAdminId)) return false
  const rollbackEvidence = actor.approvalId !== undefined || actor.requesterAdminId !== undefined
  if (rollbackEvidence) {
    return (
      actor.operationId !== undefined &&
      actor.approvalId !== undefined &&
      actor.requesterAdminId !== undefined &&
      actor.requesterAdminId !== actor.adminId
    )
  }
  return true
}

export type TokenSource = {
  token: (actor?: CloudActorContext) => string
}

export function createTokenSource(options: {
  clock?: () => Date
  issuer: string
  privateKey: KeyObject
  scopes: readonly string[]
  subject: string
}): TokenSource {
  const clock = options.clock ?? (() => new Date())
  return {
    token(actor?: CloudActorContext): string {
      if (!validActorContext(actor)) {
        throw new Error('service JWT actor context is invalid')
      }
      const now = Math.floor(clock().getTime() / 1000)
      const claims: ServiceClaims = {
        iss: options.issuer,
        sub: options.subject,
        aud: serviceJWTAudience,
        scope: [...options.scopes],
        iat: now,
        nbf: now - notBeforeSkewSeconds,
        exp: now + serviceTokenLifetimeSeconds,
        jti: randomBytes(16).toString('base64url'),
      }
      if (actor) {
        claims.admin_id = actor.adminId
        claims.admin_role = actor.role
        if (actor.operationId) claims.operation_id = actor.operationId
        if (actor.approvalId) {
          claims.approval_id = actor.approvalId
          claims.requester_admin_id = actor.requesterAdminId
        }
      }
      const header = Buffer.from(JSON.stringify({ alg: 'EdDSA', typ: 'JWT' })).toString('base64url')
      const body = Buffer.from(JSON.stringify(claims)).toString('base64url')
      const unsigned = `${header}.${body}`
      const signature = signRaw(null, Buffer.from(unsigned), options.privateKey)
      return `${unsigned}.${signature.toString('base64url')}`
    },
  }
}
